Snake game: Fix direction bug

fabionardelli opened the pull request:
There was a bug in the snake game: if the user pressed a key in a direction opposite to the snake's current direction, the game would stop and the "game over!" message would appear. The reason was that a direction change like this led to a "collision" of the snake into itself. E.g.: if the snake was running up and the user pressed the down key, the game would stop.
I added a simple control to check this, avoiding to update the direction if the new one is the opposite of the previous one.

fabionardelli committed:
added a control to avoid game over when the new key pressed is the opposite of the previous one

improved hotfix implementation
using dictionary to store opposite directions pairs

Update example8.py

Update 03. Creating a Snake Game.ipynb

nikhilkumarsingh commented:
Hey. Sorry for late reply. I would like you to do 2 more things here:
1. Squash your commits for this PR into a single commit.
2. Do the same code change in actual python script in `code` folder as well.

fabionardelli commented:
Hi. First of all I would like to thank you for accepting my PR (technically this is a closed source repo, so maybe sending PRs is not orthodox, I apologize for that, I didn't realize before). I squashed the commits into one and fixed the code in the `code` folder, hope I didn't make mistakes. Thank you again, your tutorial really inspired me!

nikhilkumarsingh commented on code/example8.py:
```diff
@@ -47,7 +47,17 @@ def main(stdscr):
 		# set direction if user pressed any arrow key
 		if key in [curses.KEY_RIGHT, curses.KEY_LEFT, curses.KEY_DOWN, curses.KEY_UP]:
-			direction = key
+			# store the previous snake's direction, in order to
+			# avoid that the new direction is the opposite of
+			# the previous one on the same axis.
+			# E.g.: the new direction can't be up/down/left/right
+			# if the previous direction was down/up/right/left
+			prev_direction = direction
+			if (key == curses.KEY_UP and prev_direction != curses.KEY_DOWN or
+		    		key == curses.KEY_DOWN and prev_direction != curses.KEY_UP or
+	            		key == curses.KEY_RIGHT and prev_direction != curses.KEY_LEFT or
+	            		key == curses.KEY_LEFT and prev_direction != curses.KEY_RIGHT):
+				direction = key
```
Let us keep a dictionary of opposite directions so that we do not have to do so many comparisons everytime.
```py
OPPOSITE_DIRECTION_DICT = {
    curses.KEY_UP: curses.KEY_DOWN,
    curses.KEY_DOWN: curses.KEY_UP,
    curses.KEY_RIGHT: curses.KEY_LEFT,
    curses.KEY_LEFT: curses.KEY_RIGHT
}

DIRECTIONS_LIST = [curses.KEY_RIGHT, curses.KEY_LEFT, curses.KEY_DOWN, curses.KEY_UP]
```

```py
if key in DIRECTIONS_LIST and key != OPPOSITE_DIRECTION_DICT[direction]:
     direction = key
```

fabionardelli commented:
Hi, I modified the code as you suggested.

nikhilkumarsingh commented:
Hey, the constants (`DIRECTIONS_LIST` and `OPPOSITE_DIRECTION_DICT`) should not be re-assigned again and again in a loop. I have put them outside. Also removed unnecessary `if` condition.

Just squash the 4 commits into one and I will merge.

fabionardelli commented:
> Hey, the constants (`DIRECTIONS_LIST` and `OPPOSITE_DIRECTION_DICT`) should not be re-assigned again and again in a loop. I have put them outside. Also removed unnecessary `if` condition.
> 
> Just squash the 4 commits into one and I will merge.

Should I squash all these commits?
![commits](https://user-images.githubusercontent.com/48182427/96371637-e1576380-1162-11eb-8158-e4bee138143b.png)

nikhilkumarsingh commented:
Yes please.

fabionardelli commented:
> Yes please.

Done.
